2.0 - Add `IDefinitionStore` specification & reference implementation + Various improvements

SakuraIsayeki opened the pull request.

SakuraIsayeki committed:
Init `Nodsoft.WowsReplaysUnpack.FileStore` library project

This project will hold the main implementation for proposal #22.

SakuraIsayeki committed:
Breakout `IDefinitionLoader` from `IDefinitionStore` + Implementations

SakuraIsayeki committed:
Update DI extensions & builder to include `IDefinitionLoader`

stewieoO committed:
Fix the DefinitionLoader Extraction

stewieoO committed:
fix visibility of loader

SakuraIsayeki committed:
Draft Filesystem defs loader implementation

SakuraIsayeki committed:
Add integration tests for unpacker configurations (default + fs)

stewieoO committed:
Fix Version Loading

stewieoO committed:
Add Comment For GetSupportedVersions()

SakuraIsayeki committed:
HACK: Disable tests parallelization

This commit mitigates issue #23, pending a more potent fix.

HenryQuan committed:
Support running build manually

HenryQuan committed:
Use nbgv 0.4.1

HenryQuan committed:
Allow testing in parallel

HenryQuan committed:
Debug multiple tasks in Program

HenryQuan committed:
Fixed issue with mutlithreading

HenryQuan committed:
Added simple benchmark in Program

SakuraIsayeki committed:
Merge pull request #29 from wowsinfo/feature/defs-filestore

Add Semaphore to address #23

HenryQuan committed:
Merge branch 'develop' into feature/defs-filestore

HenryQuan committed:
Fix missing consts

HenryQuan committed:
Update the samplePath

HenryQuan committed:
Update sameFolder in tests

HenryQuan committed:
Add a timeout to prevent dead locks

SakuraIsayeki committed:
Merge pull request #30 from Nodsoft/feature/defs-filestore

Merge feature/defs-filestore to develop

BloodyTakao committed:
feat(version/12_10_0): Add gamedefinitons and example replay

stewieoO commented on Nodsoft.WowsReplaysUnpack.FileStore/Definitions/FileSystemDefinitionLoader.cs:
```diff
@@ -0,0 +1,69 @@
+		return versions;
+	}
+	
+	[MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Why though :D
The GetSupportedVersion() method is called only once when the store is constructed.
To my knowledge the aggressive inlining is only usefull in some edgecases on hotpaths with a fuckton of calls to it

HenryQuan replied:
If it is only used once, maybe this function itself can be removed. Inline is useful for tiny functions at least for C++, not sure what C# does. Also, the compiler may make some functions Inline with optimisation turned on.

stewieoO replied:
This too yeah, the compiler already inlines by itself where it can. The attribute is just a "heavy suggestion" to the compiler to do this. It's not even guaranteed. But as i said, unless it's on a hotpath with LOTS of calls, it won't have any advantage.

stewieoO commented on Nodsoft.WowsReplaysUnpack.FileStore/Definitions/FileSystemDefinitionLoaderOptions.cs:
```diff
@@ -0,0 +1,17 @@
+/// <summary>
+/// Defines settings for the <see cref="FileSystemDefinitionLoader"/>.
+/// </summary>
+public record FileSystemDefinitionLoaderOptions
```
I know you love records, but don't you remember when we had a chat about it?
If you don't need it's special features (like equality and immutability) it actually adds more overhead than a normal class.
Not that it really matters for an options object that's only there once. But you know, gotta mention it :D

stewieoO commented on Nodsoft.WowsReplaysUnpack.Tests/Utilities.cs:
```diff
@@ -0,0 +1,23 @@
+{
+	private static readonly string _sampleFolder = Path.Join(Directory.GetCurrentDirectory(), "Replay-Samples");
+	
+	[MethodImpl(MethodImplOptions.AggressiveInlining)]
```
🤐

stewieoO commented on Nodsoft.WowsReplaysUnpack/ReplayUnpackerBuilder.cs:
```diff
@@ -0,0 +1,113 @@
+	}
+
+
+  // stewie says: No need for that since they will be added either way if you don't add other ones
```
i said what? 😆

stewieoO commented on Nodsoft.WowsReplaysUnpack/Services/ReplayUnpackerService.cs:
```diff
@@ -100,14 +106,13 @@ public UnpackedReplay Unpack(Stream stream, ReplayUnpackerOptions? options = nul
-
 		Version gameclientVersion = Version.Parse(arenaInfo.ClientVersionFromExe.Replace(',', '.'));
-		
+		_semaphore.Wait(_semephoreTimeOut);
```
I saw this was added to "prevent a deadlock" but i'm curious what resource was actually deadlocked?
And this should definitely be a configurable option.

HenryQuan replied:
I believe the deadlock still exists in this pull request when there are many threads like 12 or even more. It is also likely to happen in test. The reason behind is still unknown, I believe some functions are still holding the semephore in rare occasions.
The true root cause was that the implementation didn't consider thread safety in the first place. A rewrite may be required for this to work properly. However, it can be time consuming.

stewieoO replied:
@HenryQuan do you know what gets deadlocked though? Does it only happen when files are read from the filesystem or also when it's coming from memory? Or is it purely because it's not async? I'm fairly certain i had it async at the start but then changed it back because it made no sense somewhere...

HenryQuan replied:
The issue without semephores was due to the decryption. Please see, https://github.com/Nodsoft/WoWS-ReplaysUnpack/pull/29#issue-1921067190. I don't understand why `BinaryReader` isn't thread safe. It seems that other threads are changing its pointer also as it is like a singleton. With semephores, in a heavy multithread condition, there may be some deadlocks. Not sure what is blocking it because there shouldn't be any locks. We probably to investigate further on this. I am not very familar with C# semephore also.

stewieoO replied:
Ah, so initially the issue wasn't deadlocks but some weird errors with the BinaryReader and now with the added Semaphore there's the potential for deadlocks?

Also the issue you mentioned is about the new file system file store right?

So am i right in assuming the previous issue and the deadlocks only happen with the file system store and not when it's coming from memory?

HenryQuan replied:
That's correct I believe. Maybe, it is a lock coming from the file system. I have not tested with the memory approach yet.

stewieoO replied:
That would make more sense yeah. I wouldn't even know where to begin if it's not caused by the file system...
I just had a quick look at the code cause i couldn't quite remember but i think the issue could be that the definitions are loaded on-demand and then cached for future access.
This wasn't an issue when i first implemented it with the embedded resources i guess, but when you open them from the FS there's a potential for file locks or whatever. So maybe the filesystem loader just needs a lock(obj) thingy inside the GetFileAsXml method. it will be slower for the first access of each definition (if there's multiple threads doing it at the same time) but then they're cached anyways.

stewieoO committed:
Fix the order of arguments being prepended

Fixes #28

stewieoO commented:
@SakuraIsayeki forgot one thing, the readme needs to be updated.
The description for the `IDefinitionStore` is not correct anymore and a description for `IDefinitionLoader` needs to be added

SakuraIsayeki committed:
Merge pull request #32 from Nodsoft/bugfix/wrong-argument-order

Fix the order of arguments being prepended to method calls

SakuraIsayeki committed:
fix: Switch `IDefinitionLoader` to singleton

Switched the registration of `IDefinitionLoader` from scoped to singleton in the `ReplayUnpackerBuilder` class. This change ensures that only one instance of the definition loader is created and used throughout the application, improving performance and consistency.

SakuraIsayeki committed:
feat(tests): Add tests for ExtendedData controller

- Added a reference to the Nodsoft.WowsReplaysUnpack.ExtendedData project in the Nodsoft.WowsReplaysUnpack.Tests.csproj file.
- Added a new replay sample file, 12.10_2.wowsreplay.
- Updated ReplayUnpackerTests.cs to include tests for parsing a working replay using the ExtendedData controller.
- Registered the ExtendedData controller manually in the TestReplay_ExtendedData_Pass method.

BloodyTakao committed:
feat(version/12_11_1): Add gamedefinitions and example replay

SakuraIsayeki committed:
refactor: Various refactors

- Updated the `GetFileAsXml` method in `FileSystemDefinitionLoader.cs` to replace dots with underscores in the client version string when constructing the file path.
- Modified the `GetVersions` method in `FileSystemDefinitionLoader.cs` to create a new `Version` object by replacing underscores with dots in directory names.
- Changed the record declaration of `FileSystemDefinitionLoaderOptions` to a class declaration.

BloodyTakao committed:
feat(version/13_0_0): Add gamedefinitions and example replay

BloodyTakao committed:
feat(version/13_0_1): Add gamedefinitions and example replay

SakuraIsayeki commented:
Merging this PR now, pending follow-up on outstanding issues :
- README needs to be updated to include `IDefinitionLoader` and subsequently update `IDefinitionStore`.
- Deadlock issue should be investigated separately, yet is not a blocking issue (high risk, low impact).
